diff --git a/Sol/Model/pybullet_drone_simulator.py b/Sol/Model/pybullet_drone_simulator.py
index bbe4ae1..9d53e8a 100644
--- a/Sol/Model/pybullet_drone_simulator.py
+++ b/Sol/Model/pybullet_drone_simulator.py
@@ -261,9 +261,9 @@ class PBDroneSimulator:
         eval_env = VecNormalize(eval_env, norm_obs=True, norm_reward=True,
                                 clip_obs=1)
 
-        # onpolicy_kwargs = dict(activation_fn=th.nn.ReLU,
-        #                        net_arch=dict(vf=[512, 512, 256, 128],
-        #                                      pi=[512, 512, 256, 128]))
+        onpolicy_kwargs = dict(activation_fn=th.nn.Tanh,
+                               net_arch=dict(vf=[512, 512, 256, 128],
+                                             pi=[512, 512, 256, 128]))
         # onpolicy_kwargs = dict(net_arch=[512, 512, dict(vf=[256, 128], pi=[256, 128])])
 
         model = PPO("MlpPolicy",
@@ -278,8 +278,8 @@ class PBDroneSimulator:
                     learning_rate=1e-3,
                     tensorboard_log="./logs/ppo_tensorboard/",
                     device="auto",
-                    policy_kwargs=  # onpolicy_kwargs
-                    dict(net_arch=[256, 256, 256], activation_fn=th.nn.GELU, ),
+                    policy_kwargs=   onpolicy_kwargs
+                    # dict(net_arch=[256, 256, 256], activation_fn=th.nn.GELU, ),
                     )
 
         # tensorboard --logdir ./logs/ppo_tensorboard/
