diff --git a/Sol/Model/PBDroneEnv.py b/Sol/Model/PBDroneEnv.py
index 5575706..082d4f9 100644
--- a/Sol/Model/PBDroneEnv.py
+++ b/Sol/Model/PBDroneEnv.py
@@ -289,7 +289,7 @@ class PBDroneEnv(
         """
         if self._computeTerminated() and not self._is_done:
             # print("term and NOT DONE")
-            return -3000
+            return -300
             # -10 * (len(self._target_points) - self._current_target_index)) #  * np.linalg.norm(velocity)
 
         reward = 0.0
@@ -323,10 +323,10 @@ class PBDroneEnv(
                 # self.reward += max(3.0 * self.waypoints.progress_to_target(), 0.0)
 
                 # Add a negative reward for spinning too fast
-                reward += -np.linalg.norm(self.ang_v) / 3
+                reward += -np.linalg.norm(self.ang_v) / 30
 
                 # Penalize large actions to avoid erratic behavior
-                reward -= 0.01 * np.linalg.norm(self._last_action)
+                # reward -= 0.01 * np.linalg.norm(self._last_action)
 
         except ZeroDivisionError:
             # Give a high reward if the drone is at the target (avoiding division by zero)
diff --git a/Sol/Model/__pycache__/PBDroneEnv.cpython-38.pyc b/Sol/Model/__pycache__/PBDroneEnv.cpython-38.pyc
index f50a774..a670134 100644
Binary files a/Sol/Model/__pycache__/PBDroneEnv.cpython-38.pyc and b/Sol/Model/__pycache__/PBDroneEnv.cpython-38.pyc differ
diff --git a/Sol/Model/pybullet_drone_simulator.py b/Sol/Model/pybullet_drone_simulator.py
index ef225dd..e1adf56 100644
--- a/Sol/Model/pybullet_drone_simulator.py
+++ b/Sol/Model/pybullet_drone_simulator.py
@@ -343,7 +343,7 @@ class PBDroneSimulator:
                         train_env,
                         verbose=1,
                         n_steps=2048 * self.num_envs,
-                        batch_size=49152 // self.num_envs,
+                        batch_size=49152 * self.num_envs,
                         ent_coef=0.01,
                         # use_sde=True,
                         # sde_sample_freq=4,
